### PR TITLE
Partially avoid needing QDox by using -parameters

### PR DESCRIPTION
**Pull request by octylFractal**
This only works for Java 8, but it is necessary to avoid issues with using `TYPE_USE` annotations. This change improves debugging Gradle as well, since the parameter info is available at all times in Java 8 code.

### CODE REVIEW

File: build-logic/jvm/src/main/kotlin/gradlebuild/jvm/extension/UnitTestAndCompileExtension.kt [hunk 1]
```diff
@@ -50,8 +54,11 @@ abstract class UnitTestAndCompileExtension(private val tasks: TaskContainer) {
     fun enforceJava6Compatibility() {
         tasks.withType<JavaCompile>().configureEach {
             options.release.set(null)
+            options.compilerArgs.remove("-parameters")
             sourceCompatibility = "6"
             targetCompatibility = "6"
         }
+        // Apply ParameterNamesIndex since 6 doesn't support -parameters
+        project.apply(plugin = "gradlebuild.api-parameter-names-index")
```
Comment (big-guy) [inline build-logic/jvm/src/main/kotlin/gradlebuild/jvm/extension/UnitTestAndCompileExtension.kt:62]: are there a small enough number of projects that we could apply this to them specifically? it's a little ugly to apply a plugin from a method call like this.

**Comment by big-guy** [comment]
@eskatos Octavia is looking to do this because we want to add `@Nullable` to some type parameters, but for reasons that she can explain, QDox doesn't support that and so we're trying to remove QDox and just use Java's `-parameters`.  Is this something we can do or is there some gotcha that we still need QDox for Kotlin DSL?

**Comment by eskatos** [comment]
The data generated by `api-parameter-names-index`  is used to generate the Kotlin DSL extension jar for the Gradle API.
The process should fallback to this data if no parameter names are available in the bytecode so IIRC it should be fine to replace this with `-parameters` for projects where it is possible.

There are unit and integration tests for this so if it breaks you should get some feedback.

We should keep it for java 6 projects that contain public API stuff though. I'm pretty sure there isn't test coverage specific to Java 6 only types.

File: build-logic/jvm/src/main/kotlin/gradlebuild/jvm/extension/UnitTestAndCompileExtension.kt (hunk 1 above)
Comment (octylFractal) [reply build-logic/jvm/src/main/kotlin/gradlebuild/jvm/extension/UnitTestAndCompileExtension.kt:62]: There are 23 projects, it's do-able but maybe a little much, I would rather keep this consistent as it should be applied to any project that has Java 6 applied.

**Comment by big-guy** [comment]
@bot-gradle test this

**Comment by octylFractal** [comment]
Seems like I made the distribution too large? Should I just adjust this test (`distribution size should not exceed a certain number`)?

**Comment by octylFractal** [comment]
> What about Java 6 modules?

I left those using the `api-parameternames-index`. Some tests do cover them, and it works just fine still.

**Comment by octylFractal** [comment]
@bot-gradle test this

File: build-logic/jvm/src/main/kotlin/gradlebuild/jvm/extension/UnitTestAndCompileExtension.kt (hunk 1 above)
Comment (eskatos) [inline build-logic/jvm/src/main/kotlin/gradlebuild/jvm/extension/UnitTestAndCompileExtension.kt:62]: :+1:

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/codegen/ApiTypeProvider.kt [hunk 2]
```diff
@@ -134,7 +133,7 @@ class ApiTypeProvider internal constructor(
     private
     fun classNodeFor(classBytesSupplier: () -> ByteArray) = {
         ApiTypeClassNode().also {
-            ClassReader(classBytesSupplier()).accept(it, SKIP_DEBUG or SKIP_CODE or SKIP_FRAMES)
+            ClassReader(classBytesSupplier()).accept(it, SKIP_CODE or SKIP_FRAMES)
```
Comment (eskatos) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/codegen/ApiTypeProvider.kt:136]: :+1:

**Review by eskatos — APPROVED** [review]
It's great we get to this :tada:

**Review by big-guy — APPROVED** [review]

**Comment by big-guy** [comment]
@bot-gradle test and merge

**Pull request merged**